dev/doc/proof-engine.md

```diff
@@ -0,0 +1,130 @@
+Tutorial on the new proof engine for ML tactic writers
+======================================================
+
+Starting from Coq 8.5, a new proof engine has been introduced, replacing the old
+meta-based engine which had a lot of both expressivity and soundness drawbacks.
+The latter is deprecated and is slowly getting removed from the source code.
+
+The new engine relies on a monadic API defined in the `Proofview` module. Helper
+functions and higher-level operations are defined in the `Tacmach` and
+`Tacticals` modules, and end-user tactics are defined amongst other in the
+`Tactics` module.
+
+At the root of the engine is a representation of proofs as partial terms that
+can contain typed holes, called evars. An evar is essentially defined by its
+context and return type, that we will write `?e : [Γ ⊢ _ : A]`. An evar `?e`
+must be applied to a substitution `σ` of type `Γ` (i.e. a list of terms) to
+produce a term of type `A`, which is done by applying `mkEvar`, and which we
```

Comment:
Then, how about writing the doc as if the EConstr branch were merged. It's not particularly controversial, is it?